File: Army(c++)/State/WereWolfState.cpp
#include "WereWolfState.hpp"
#include "WolfState.hpp"
#include "../Basic/Unit.hpp"

WereWolfState::WereWolfState(Unit* thisUnit) : thisUnit(thisUnit) {}

void WereWolfState::changeUnit(Unit* enemy) {
    this->thisUnit->getHealth().ensureIsAlive();
    enemy->getHealth().ensureIsAlive();
    
    if ( enemy->getWarriorID() == VAMPIRE || enemy->getWarriorID() == WEREWOLF || enemy->getWarriorID() == WOLF || enemy->getWarriorID() == DEMON ) {
        return;
    }
    
    enemy->transitionTo(new WereWolfState(enemy));
    
    enemy->setWarriorID(WarriorID::WEREWOLF);
    enemy->getHealth().setDamage(this->thisUnit->getHealth().getDamage());
    enemy->getHealth().setHitPoints(this->thisUnit->getHealth().getHitPointsLimit());
    enemy->getHealth().setHitPointsLimit(this->thisUnit->getHealth().getHitPointsLimit());
}

void WereWolfState::transformTo() {
    this->thisUnit->getHealth().ensureIsAlive();
    
    this->thisUnit->transitionTo(new WolfState(this->thisUnit));
    
    this->thisUnit->setWarriorID(WarriorID::WOLF);
    this->thisUnit->getHealth().setDamage(this->thisUnit->getHealth().getDamage()*1.5);
    this->thisUnit->getHealth().setHitPoints(this->thisUnit->getHealth().getHitPoints()*2);
    this->thisUnit->getHealth().setHitPointsLimit(this->thisUnit->getHealth().getHitPointsLimit()*2);
}

File: Army(java)/State/WereWolfState.java
package State;
import Basic.*;
import EnumID.*;

public class WereWolfState implements State {
    protected Unit thisUnit;

    public WereWolfState(Unit unit) {
        this.thisUnit = unit;
    }

    public void printState() {
        System.out.println("\tI'm WereWolf!!!");
    }

    public void changeUnit(Unit enemy) throws UnitIsDeadException {
        thisUnit.getHealth().ensureIsAlive();
        enemy.getHealth().ensureIsAlive();
        
        if ( enemy.getWarriorID() == WarriorID.VAMPIRE || enemy.getWarriorID() == WarriorID.WEREWOLF || enemy.getWarriorID() == WarriorID.WOLF || enemy.getWarriorID() == WarriorID.DEMON ) {
            System.out.println("Error!!!");
            return;
        }

        enemy.setState(new WereWolfState(enemy));
        
        enemy.setWarriorID(WarriorID.WEREWOLF);
        enemy.getHealth().setDamage(thisUnit.getHealth().getDamage());
        enemy.getHealth().setHitPoints(thisUnit.getHealth().getHitPointsLimit());
        enemy.getHealth().setHitPointsLimit(thisUnit.getHealth().getHitPointsLimit());
    }

    public void transformTo() throws UnitIsDeadException {
        thisUnit.getHealth().ensureIsAlive();
        
        thisUnit.setState(new WolfState(thisUnit));
        
        thisUnit.setWarriorID(WarriorID.WOLF);
        thisUnit.getHealth().setDamage(thisUnit.getHealth().getDamage()*2);
        thisUnit.getHealth().setHitPoints(thisUnit.getHealth().getHitPoints()*2);
        thisUnit.getHealth().setHitPointsLimit(thisUnit.getHealth().getHitPointsLimit()*2);
    }
}
